### DISCUSSION TITLE
Perform inline does not handle retries

### DISCUSSION THREAD

**Discussion by linas-deliveroo** [sidekiq/sidekiq#5656] [General]
Hi!

Thank you for Sidekiq! It's central to many things that I do!

I have a network call that fails rarely. Obviously, it should run in a job. However, most of the time it succeeds, and in that case I want to find out about it synchronously (say, in the same web request).

`perform_sync` sounded exactly like what I needed! However, it's not a fully fledged way to run a job:
- it does not handle exceptions nor schedule retries,
- does not update stats,
- does not log progress.

`Processor#process` does all of these, in particular `Processor#dispatch` would be super handy. However, they are both private.

Any ideas how I could run a job inline once and let it retry upon failure?

Thank you!

**Comment by mperham** [comment]
I'm not going to support inline execution which looks like async execution. It's not going to call middleware, update stats or retry because it's not executing in Sidekiq. But you can always build something like it:

```ruby
def inline_with_backup(klass, *args)
  begin
    klass.perform_sync(*args)
  rescue
    klass.perform_in(1.minute, *args)
  end
end

inline_with_backup(FooJob, 1, "mike")
```

**Reply by linas-deliveroo** [reply]
Thank you for your reply!

Your example is something I thought I'd go with initially. However, it messes with error states and retry counts so in case anyone is interested something like this seems to also work (so far) instead:
```ruby
def self.perform_inline_with_retries(*args)
  queue = get_sidekiq_options.fetch('queue')
  job_hash = normalize_item('queue' => queue, 'class' => name, 'args' => args)
  jobstr = JSON.dump(job_hash)
  retrier = Sidekiq::JobRetry.new
  begin
    retrier.global(jobstr, queue) do
      retrier.local(new, jobstr, queue) do
        perform_inline(*args)
      end
    end
  rescue Sidekiq::JobRetry::Handled
    # Job failed and will be retried.
  end
end
```
Of course, most of this is private Sidekiq machinery so YMMV… Thank you!
